File: Trackify_Admin/src/lib/services/missionLogger.service.js
import { db } from '../firebase';
import { collection, addDoc, serverTimestamp } from 'firebase/firestore';

export const MissionSeverity = {
  INFO: 'INFO',
  SUCCESS: 'SUCCESS',
  WARNING: 'WARNING',
  ERROR: 'ERROR',
  CRITICAL: 'CRITICAL',
  SECURITY: 'SECURITY'
};

export const MissionModule = {
  Attendance: 'Attendance',
  Payroll: 'Payroll',
  Reports: 'Reports',
  Subscription: 'Subscription',
  Security: 'Security',
  Firebase: 'Firebase',
  TrackOps: 'TrackOps',
  Errors: 'Errors',
  Contractors: 'Contractors',
  Labours: 'Labours',
  Critical: 'Critical'
};

/**
 * Centralized service to push audit logs to the Trackify Mission Control stream.
 */
class MissionLoggerService {
  /**
   * Log an action to the `mission_logs` collection.
   */
  async logAction({ severity, module, action, companyId, userId, details }) {
    try {
      const payload = {
        timestamp: serverTimestamp(),
        severity: severity,
        module: module,
        action: action,
        companyId: companyId,
        userId: userId,
        details: details || ''
      };

      await addDoc(collection(db, 'mission_logs'), payload);
    } catch (err) {
      console.warn('Failed to send mission log:', err);
    }
  }
}

export default new MissionLoggerService();

File: lib/services/mission_logger_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';

enum MissionSeverity { Info, Success, Warning, Error, Critical }
enum MissionModule { UserLogs, BusinessLogs, SecurityLogs, ProductLogs, SystemLogs }

/// Centralized service to push audit logs to the Trackify Mission Control stream.
class MissionLoggerService {
  MissionLoggerService._();
  static final MissionLoggerService instance = MissionLoggerService._();

  final FirebaseFirestore _db = FirebaseFirestore.instance;

  /// Log an action to the `mission_logs` collection.
  Future<void> logAction({
    required MissionSeverity severity,
    required MissionModule module,
    required String action,
    required String companyId,
    required String userId,
    required String role,
    String? details,
  }) async {
    try {
      final payload = {
        'timestamp': FieldValue.serverTimestamp(),
        'severity': severity.name,
        'module': module.name,
        'action': action,
        'companyId': companyId,
        'userId': userId,
        'role': role,
        'details': details ?? '',
      };

      await _db.collection('mission_logs').add(payload);
    } catch (e) {
      debugPrint('Failed to send mission log: $e');
    }
  }
}
